DSEGOG-128 Image centroids/crosshairs

louise-davies opened the pull request:
## Description
Adds the ability to go into "crosshairs" mode. When switching into this mode, it will display the centroid for the image and the corresponding intensity plots. The user can click on the image to reposition the crosshairs. Resetting the view resets back to the centroid.

It should work when switching via the thumbnail selector, we always default to the centroid though.

![image](https://github.com/user-attachments/assets/cead5612-5a85-4b0c-9a09-74dbed5673b9)


## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] Play around with the crosshairs mode

## Agile board tracking
Closes DSEGOG-128

louise-davies committed:
DSEGOG-128 add crosshair to image window
Just basic API hookup & crosshair drawing for now
Had to refactor some of the canvas code to reduce blur

louise-davies committed:
DSEGOG-128 add Y axis intensity plot
- Had to slightly change the chart.js code to be able to load multiple plots in a window

louise-davies committed:
DSEGOG-128 add x plot
- Also make it so the plot is sized correctly when the image loads

louise-davies committed:
DSEGOG-128 fix issues with crosshairs not being drawn, fix reset behaviour & hide plots when being redrawn

louise-davies committed:
DSEGOG-128 display position and FWHMs, plus some fixing the scales code to only be set when the image size is known

louise-davies committed:
DSEGOG-128 change image control panel position when intensity graphs are displayed

louise-davies committed:
DSEGOG-128 update snapshots

louise-davies committed:
DSEGOG-128 fix zoom tests broken by refactor & add imagePlot tests

louise-davies committed:
DSEGOG-128 add tests for crosshairs mode to imageView

louise-davies committed:
DSEGOG-128 add crosshair tests for imageControlsPanel, imageWindow & useImageCrosshairs

louise-davies committed:
Merge branch 'develop' into DSEGOG-128-image-centroids

louise-davies committed:
DSEGOG-128 update chart selector in existing e2e tests

louise-davies committed:
DSEGOG-128 add mocked e2e tests for crosshairs

louise-davies committed:
DSEGOG-128 update snapshots

louise-davies committed:
DSEGOG-128 increase stability of image e2e tests
- fix some assertions which weren't actually retrying due to await in wrong place
- add some extra assertions to help checking if images change

louise-davies committed:
DSEGOG-128 add "real" e2e test case for crosshairs

joshdimanteto commented on src/images/imagePlot.component.tsx:
```diff
@@ -0,0 +1,213 @@
+    JSON.stringify(
+      type === 'x'
+        ? {
+            width: imageElement?.naturalWidth
+              ? imageElement.naturalWidth + 29
+              : 200,
+            height: 200,
+          }
+        : {
+            width: 200,
+            height: imageElement?.naturalHeight
+              ? imageElement.naturalHeight + 22
+              : 200,
```
Why does the natural width and height need 29 and 22 added ?

louise-davies replied:
Ah, I should add a comment for this, but I needed to add some padding to account for the width/height of the Chart.js axis ticks (aka the numbers) so that the actual axis line itself aligns with the image.

Maybe I can try and calculate it based on the default config variables in Chart.js - but they may just have to remain magic numbers...

I will at least extract both out to variables at the top of the file (to make them easy to edit if we need) and explain what they're for

joshdimanteto reviewed (dismissed):
LGTM, work well just one question.

joshdimanteto commented:
Should an exported images also include the crosshairs ??

louise-davies commented:
@joshuadkitenge no it shouldn't (I think - they're not exported in eCat) - if they do it's an oversight on my part (although I wouldn't think they would as I would have thought we download the raw image itself, not the canvas overlay where the crosshairs are drawn...)

louise-davies committed:
Merge branch 'develop' into DSEGOG-128-image-centroids

louise-davies committed:
DSEGOG-128 extract intensity plot offsets to variables & add comments

louise-davies committed:
DSEGOG-128 improve intensity plot sizing
- this helps reduce blurriness when moving between displays with different devicePixelRatio

louise-davies committed:
DSEGOG-128 calculate plot axis offsets & refactor image plot to be more reliable after the responsive: true change

louise-davies committed:
DSEGOG-128 revert calculation of offset changes & fix e2e tests
- the calculation stuff was unreliable across browsers, instead just use the constants we know currently work
and rely on our snapshot testing to highlight any issues if defaults change in the future

louise-davies committed:
Merge branch 'develop' into DSEGOG-128-image-centroids

louise-davies committed:
DSEGOG-128 use dynamic range for intensity plots
- fix alignment issues caused by difference between 16-bit & 8-bit images

louise-davies committed:
DSEGOG-128 update tests due to changes to accommodate 16-bit images intensity plots

louise-davies commented:
@joshuadkitenge this is ready for review again as Stephen got back to me on how to handle 16-bit images. He picked the option where we let our plotting library autoscale the plot. See below example of a 16-bit image with intensity plots activated

![image](https://github.com/user-attachments/assets/2a2e77cf-91e2-4cf1-b902-77cc1c6489ac)

louise-davies committed:
DSEGOG-128 fix copy-paste error & update snapshot

joshdimanteto reviewed (commented):
LGTM to one minor thing. The plot on the right-hand side could do with a bit more spacing from the image so it either matches the spacing of the image underneath or is inline with the image control panel. 

![image](https://github.com/user-attachments/assets/c0bb5253-6e0b-473e-ad00-85a5c6048e63)

louise-davies commented:
@joshuadkitenge I forgot to add spacing to the corresponding grid for the top row. I can't get it to align perfectly with the control panel (as the horizontal spacing of the right-hand graph kind of depends on the scale of the image pixel axis) but it aligns with the numbers on the other graph now

![image](https://github.com/user-attachments/assets/2bd69a1f-46b4-4b7d-a9de-f1c95bf7277f)

joshdimanteto commented:
That looks good.

louise-davies committed:
DSEGOG-128 fix spacing in image window
